tool/metrics.py

```diff
@@ -0,0 +1,113 @@
+"""
+Implements the Distinct-n metric for evaluating diversity in dialogues.
+The implementation is copied from Joko et al. (2024) which is based on Tevet et al. (2021).
+For more details, refer to Joko et al. (2024).
+
+References:
+- Joko et al. (2024): https://dl.acm.org/doi/10.1145/3626772.3657815
+    (GitHub: https://github.com/informagi/laps)
+- Tevet et al. (2021): https://aclanthology.org/2021.eacl-main.25/
+"""
+
+import random
+from nltk.tokenize import word_tokenize
+import numpy as np
+from typing import List, Tuple, Union
+
+def _truncate_response_set(response_set: List[str], max_token_length: int) -> List[str]:
+    truncated_response_set = []
+    token_count = 0
+    for line in response_set:
+        words = line.split()
+        if token_count + len(words) <= max_token_length:
+            truncated_response_set.append(line)
+            token_count += len(words)
+        else:
+            remaining_space = max_token_length - token_count
+            words_to_append = words[:remaining_space]
+            truncated_response_set.append(' '.join(words_to_append))
+            token_count += len(words_to_append)
+            break
+    
+    # Check if total token count is less than max_token_length
+    if token_count < max_token_length:
+        print("\033[93mWarning: The response did not reach the maximum token limit.\033[0m")
+
+    return truncated_response_set
+
+class DistinctNgrams:
+    """Class to calculate Distinct-n metric as per Tevet's implementation."""
+
+    def __init__(self, n: int = 2, tokenizer_name: str = 'split', use_lower_case: bool = True) -> None:
+        self.n = n
+        self.tokenizer_name = tokenizer_name
+        self.use_lower_case = use_lower_case
+
+    def _tokenize(self, line: str) -> List[str]:
+        """Tokenizes the input line based on the specified tokenizer_name.
+
+        Args:
+        line (str): The input line to _tokenize.
+
+        Returns:
+        list: A list of tokens from the input line.
```

Comment:
Nit: add indentation for consistency.

```suggestion
        Args:
            line (str): The input line to _tokenize.

        Returns:
            list: A list of tokens from the input line.
```